README.md
# Twill Image

Twill Image is a package designed to work with [Twill](https://twill.io) to display responsive images easily on your site. It leverages Twill's image processing capabilities and adds modern lazy-loading techniques. It supports responsive images, art direction and fixed width images.

- `<picture>` with multiple `<source>` elements
- Twill's low-quality placeholder (LQIP)
- Background-color placeholder
- Art direction (multiple crops)
- WebP and JPEG support
- Lazy load (fade in) image with `IntersectionOserver`
- Support native lazy loading with `loading='lazy'`

## Contents

- [Installation](#installation)
  - [Configuration file](#configuration-file)
  - [JavaScript module](#javascript-module)
- [Usage](#usage)
  - [The `Image` model](#the-image-,model)
    - [Available methods](#available-methods)
      - [`crop`](#crop)
      - [`width`](#width)
      - [`height`](#height)
      - [`sources`](#sources)
      - [`sizes`](#sizes)
      - [`columns`](#columns)
      - [`srcSetWidths`](#srcSetWidths)
      - [`preset`](#preset)
      - [`render`](#render)
      - [`toArray`](#toArray)
  - [The facade `render` method](#the-facade-render-method)
    - [List of arguments](#list-of-arguments)
    - [Examples](#examples)
- [Configuration](#configuration)
  - [Presets](#presets)
  - [List of options](#list-of-options)
- [Frontend breakpoints and grid structure](#frontend-breakpoints-and-grid-structure)
  - [`columns` example](#columns-example)
    - [`columns` preset](#columns-preset)
    - [`columns` output](#columns-output)
  - [`columns` custom class](#columns-custom-class)
- [Art directed images](#art-directed-images)
- [Multiple medias](#multiple-medias)

## Installation

Install the package to your existing Twill project with Composer.

```
composer require area17/twill-image
```

### Configuration file

The configuration file contains a few general settings and this is where you can define preset for your images.

Publish `config/twill-image.php` to your app's config folder.

```bash
php artisan vendor:publish --provider="A17\Twill\Image\TwillImageServiceProvider" --tag=config
```

### JavaScript module

You can import the JavaScript module and initialize the lazy loading class in your application.

```js
import { TwillImage } from '../../vendor/area17/twill-image'

document.addEventListener('DOMContentLoaded', function () {
  const lazyloading = new TwillImage()
})
```

When adding or refreshing content of a page without a reload, you can trigger a recalculation of TwillImage's observers by calling the `reset()` method. This is an example:

```js
document.addEventListener('page:updated', () => lazyloading.reset());
```

If you prefer to use a pre-compiled version of the JavaScrip module, you can publish a script `twill-image.js` to your app's public folder and add a `<script>` tag to your project.

```bash
php artisan vendor:publish --provider="A17\Twill\Image\TwillImageServiceProvider" --tag=js
```

In a Blade file.

```php
<script src="{{ asset('/twill-image.js') }}"></script>
```

The JavaScript module is not required. If you prefer to rely only on the browser's native `loading` attribute, set the [`js` config option](#list-of-options) to `false`.


## Usage

### The `Image` model

The `Image` model allows you to interact fluently with a media object.

```php
$image = new A17\Twill\Image\Models\Image($object, $role, $media);

// or using the Facade
$image = TwillImage::make($object, $role, $media);
```

|Argument|Type|Default|Description|
|---|---|---|---|
|`$object` (Required)|`A17\Twill\Models\Media` `A17\Twill\Models\Block` `object`|   |Your Twill module or block object|
|`$role` (Required)|`string`|   |`Media` role|
|`$media`|`A17\Twill\Models\Media`|`null`|`Media` instance|

#### Available methods

Once you have created an instance of the `Image` model, you can interact by using one or chaining many of these methods.
##### `crop`

You can specify the crop name by passing it as an argument. By default, the `Image` model will look for a crop name `default` and if it isn't availble, it will look for a single crop and select it. If it can't determine the crop, it will result in an error.

```php
$image->crop('listing_card');
```

##### `width`

To set the width of the image, you can use this method. The default is an image of 1000 pixels wide. This is useful if you need to display an image with a fixed width or if you know in advance that you will a larger image than the default.

Note: the width is applied to the "fallback" image (`<img src="{{ $image }}">`) and to determine the number of image URLs to add to the `srcset` attribute.

```php
$image->width(1500);
```

##### `height`

You can set the height of the image with this method. Similar to the `width` method above, it is most useful for fixed-size image. When not used, the height is determined by the aspect ratio of the image and inferred from the width.

```php
$image->height(900);

$image->crop('listing')->width(600)->height(400);
```

##### `sources`

To use mutliples `<source>` elements, you can pass a array to this method by listing the sources other than the main crop. Each item in the array must have a `mediaQuery` and a `crop` key in order to generate the proper `srcset`. You can pass an optional width and height. This is useful when used with the `crop` method to set the main image crop. See also [Art directed images](#art-directed-images).

```php
$image->crop('desktop')->sources([
    [
        'mediaQuery' => '(max-width: 400px)', // required
        'crop' => 'mobile', // required
        'width' => 200, // optional
        'height' => 200, // optional
        'srcSetWidths' => [100, 200, 400], // optional
    ],
    [
        'mediaQuery' => '(min-width: 401px) and (max-width: 700px)',
        'crop' => 'tablet',
    ],
]);
```

Media queries can also be generated from a [frontend breakpoints and grid structure](#frontend-breakpoints-and-grid-structure) file by passing a `columns` key instead of `mediaQuery`. You can see the format below.


```php
$image->crop('desktop')->sources([
    [
        'columns' => [
            'md' => 'max',
        ],
        'crop' => 'mobile',
    ],
    [
        'columns' => [
            'md' => 'min',
            'lg' => 'max',
        ],
        'crop' => 'tablet',
    ],
]);
```

##### `sizes`

Use this method to pass a `sizes` attribute to the model.

```php
$image->sizes('(max-width: 400px) 100vw, 50vw');
```

##### `columns`

As an alternative to the `sizes` method, Twill Image provides a way to generate the `sizes` attribute based on a [frontend breakpoints and grid structure](#frontend-breakpoints-and-grid-structure) file. When placing this JSON file at the base folder of your app, the `sizes` attribute can be generated from passing a series of breakpoints and columns number to this method.

```php
$image->columns([
    'xxl' => 6,
    'xl' => 6,
    'lg' => 8,
    'md' => 8,
    'sm' => 8,
    'xs' => 12,
]);
```

This would tell how many columns the image will take at each breakpoint in order to generate to proper `sizes` attribute.

Note: this method will have an effect only when `frontend.config.json` exists at in base folder of your app.

##### `srcSetWidths`

Use this method to give a list a widths to generate the `srcset` attribute. Without this method, Twill Image will auto generate a series of widths based on the image width.

```php
$image->srcSetWidths([100, 150, 300, 600, 1200, 2000, 2400, 3600, 5000]);
```

##### `preset`

With this method you can use an object to pass a value to any of the above methods. You can also add a preset key to the config `config/twill-image.php` and pass the name to this method.

```php
// config/twill-image.php

return [
    // ...
    'presets' => [
        'art_directed' => [
            'crop' => 'desktop',
            'width' => 700,
            'sizes' => '(max-width: 767px) 25vw, (min-width: 767px) and (max-width: 1023px) 50vw, 33vw',
            'srcSetWidths' => [350, 700, 1400],
            'sources' => [
                [
                    'crop' => 'mobile',
                    'media_query' => '(max-width: 767px)',
                    'srcSetWidths' => [100, 200, 400],
                ],
                [
                    'crop' => 'tablet',
                    'media_query' => '(min-width: 767px) and (max-width: 1023px)',
                ],
            ],
        ],
    ],
];
```

```php
// to use this preset from the config file
$image->preset('art_directed');
```

You can directly pass the full object if you prefer.

```php
$image->preset([
    'crop' => 'desktop',
    'width' => 700,
    'sizes' => '(max-width: 767px) 100vw, (min-width: 767px) and (max-width: 1023px) 50vw, 33vw',
    'sources' => [
        [
            'crop' => 'mobile',
            'media_query' => '(max-width: 767px)',
        ],
        [
            'crop' => 'tablet',
            'media_query' => '(min-width: 767px) and (max-width: 1023px)',
        ],
    ],
]);
```

##### `render`

This method will return the rendered view.

```blade
{{-- resources/views/home.blade.php --}}
@php
$image = TwillImage::make($page, 'preview')->preset('art_directed');
@endphp

{!! $image->render() !!}

{{-- with arguments --}}
{!! $image->render([
    'loading' => 'eager',
    'layout' => 'constrained',
]) !!}
```

##### `toArray`

If you need to split the image generation from the render (exposing the `Image` model data through a REST API for example), use this method to get all attributes as an array.

```php
$previewImage = TwillImage::make($page, 'preview')->preset('art_directed')->toArray();
```

And use the `render` method from the facade to render the view.

```blade
{{-- resources/views/page.blade.php --}}

<div>{!! TwillImage::render($previewImage) !!}</div>
```

### The facade `render` method

As seen in the previous section, the image element rendering can be separated from the image attributes generation. You can use the `Image` model to set up your image and pass the resulting object (or its `array` format to the `render` method to output the view).

```php
$previewImage = TwillImage::make($page, 'preview')->toArray();
```

```blade
{!! TwillImage::render($previewImage) !!}
```

or

```php
<div class="w-1/4">
    {!! TwillImage::render($previewImage, [
        'layout' => 'constrained',
        'width' => 700,
    ]) !!}
</div>
```

#### List of arguments

|Argument|Type|Default|Description|
|---|---|---|---|
|`backgroundColor`|`hex`  `string`|See config|Set placeholder background color|
|`class`|`string`|   |Add class(es) to the wrapper element|
|`height`|`int`|   |   |
|`layout`|`"fullWidth" \| "constrained" \| "fixed"`|`fullWidth`|By default, the image will spread the full width of its container element, `constrained` will apply a `max-width` and `fixed` will apply hard width and height value|
|`loading`|`"lazy" \| "eager"`|`lazy`|Set native lazy loading attribute|
|`lqip`|`boolean`|See config|Use LQIP|
|`sizes`|`string`|   |The image sizes attributes|
|`width`|`int`|`1000`|Used with `layout` `constrained` and `fixed`|
|`imageStyles`|`array`|`[]`|Apply styles to placeholder and main `img` tags (ex.: `[['object-fit' => 'contain']]`|

#### Examples

```blade
{!! TwillImage::make($item, 'preview_image')->sizes('(max-width: 767px) 50vw, 100vw')->render(); !!}

@php
$heroImage = TwillImage::make($item, 'preview_image');
$listingImage = TwillImage::make($item, 'preview_image')->crop('listing');
@endphp

{!! TwillImage::render($heroImage) !!}

{!! TwillImage::render($listingImage) !!}

{!! TwillImage::render($listingImage, [
    'layout' => 'constrained',
    'width' => 400
]) !!}

{!! TwillImage::render($listingImage, [
    'layout' => 'fixed',
    'width' => 100,
    'height' => 150
]) !!}
```

## Configuration


In `config/twill-image.php`, you can define general options and image presets. A preset informs the `Image::preset` method which crop to output along other options like responsive sources.

```php
<?php

return [

    'background_color' => '#e3e3e3',

    'lqip' => true,

    'webp_support' => true,

    'presets' => [
        'listing' => [
            'crop' => 'card',
            'width' => 500,
            'sizes' => '25vw',
        ],
        // ...
    ],
];
```

### Presets

See [above section](#preset) about the `preset` method.
### List of options

|Argument|Type|Default|Description|
|---|---|---|---|
|`background_color`|`string`|`#e3e3e3`|   |
|`lqip`|`boolean`|`true`|Uses Twill LQIP method to generate responsive placeholder|
|`webp_support`|`boolean`|`true`|If set to `false`, the `type` attribute is omitted from `<source>` elements|
|`js`|`boolean`|`true`|If set to `false`, lazy-loading will simply rely on the image's `loading` attribute|
|`presets`|`object`|   |   |

## Art directed images

To use different crops with media queries, you need to list the other sources in a `Image::preset` or by passing them to the `Image::sources` method. The rendered image element with have only the ratio of the main crop and other ratio need to be added with CSS.

Let's say this is your preset `art_directed` in your config:

```php
    // ...
    'art_directed' => [
        'crop' => 'desktop',
        'sizes' => '(max-width: 767px) 100vw, 50vw',
        'sources' => [
            [
                'crop' => 'mobile',
                'media_query' => '(max-width: 767px)',
            ]
        ],
    ],
    // ...
```

```blade
@php
$image = TwillImage::make($page, 'preview')->preset('art_directed');
@endphp

<div>
    {!! TwillImage::render($image, [
        'class' => 'art-directed'
    ]) !!}
</div>
```

It will output the image element with the class applied to the container.

```html
<div class="twill-image-wrapper art-directed">...</div>
```

Define styles for each breakpoint.

```css
.art-directed {
  aspect-ratio: 16 / 9;
}

@media screen and (max-width: 767px) {
  .art-directed {
    aspect-ratio: unset;
  }
}
```

## Frontend breakpoints and grid structure

We provide a way to generate `sizes` and `media` attributes by describing the structure of your page in a JSON file `frontend.config.json` placed at the base of your app. An example is provided [`frontend.config.json.example`](frontend.config.json.example).

This file describes the breakpoints, main container widths, number of columns per breakpoints and inner/outer gutters. When this file exists in your project, you can use the `columns` method on the `Image` model or the `columns` key in your preset and sources objects in order to generate dynamically the `sizes` and `media` attributes.

### `columns` example

This example assumes that you have the provided `frontend.config.json` in your app's `base_path`.

#### `columns` preset

```php
// config/twill-image.php

return [
    // ...
    'presets' => [
        'art_directed' => [
            'crop' => 'desktop',
            'columns' => [
                'xxl' => 6,
                'xl' => 6,
                'lg' => 8,
                'md' => 8,
                'sm' => 8,
                'xs' => 12,
            ],
            'sources' => [
                [
                    'crop' => 'mobile',
                    'columns' => [
                        'md' => 'max',
                    ],
                ],
                [
                    'crop' => 'tablet',
                    'columns' => [
                        'md' => 'min',
                        'lg' => 'max',
                    ],
                ],
            ],
        ],
    ],
];
```

#### `columns` output

```blade
{{-- to use this preset from the config file and render the image --}}
{!! $image->preset('art_directed')->render() !!}
```

The image source and fallback would have these `sizes` and `media` attributes (the elements have been simplified for clarity):

```html
<picture>
    <source media="(max-width: 767px)" sizes="..." srcset="...">
    <source media="(min-width: 768px) and (max-width: 1023px)" sizes="..." srcset="...">
    <img sizes="
        (max-width: 543px) calc((((100vw - 260px) / 12) * 12) + 220px),
        (min-width: 544px) and (max-width: 767px) calc((((100vw - 312px) / 12) * 8) + 168px),
        (min-width: 768px) and (max-width: 1023px) calc((((100vw - 416px) / 12) * 8) + 224px),
        (min-width: 1024px) and (max-width: 1279px) calc((((100vw - 624px) / 12) * 8) + 336px),
        (min-width: 1280px) and (max-width: 1680px) calc((((100vw - 832px) / 12) * 6) + 320px),
        761px
    " src="...">
</picture>
```

### `columns` custom class

You can provide your own custom class to be used instead of the one provided. You can create your own service and provide the class name in the config file:

```php
// config/twill-image.php

    // ...
    // default to: A17\Twill\Image\Services\ImageColumns::class
   'columns_class' => MyApp\Services\MyOwnImageColumnsService::class,

];
```

The service must implement the interface `A17\Twill\Image\Services\Interfaces\ImageColumns`.


This can also be useful if you simply need to override some of the proprties that are defined in the provided services.

## Multiple medias

This is an example when you have multiple medias attached to a single `role`.

```blade
@php
$galleryImages = $item->imageObjects('gallery_image', 'desktop')->map(function ($media) use ($item) {
    return TwillImage::make($item, 'gallery_image', $media);
})->toArray();
@endphp

@if($galleryImages)
    @foreach($galleryImages as $image)
        {!! TwillImage::render($image) !!}
    @endforeach
@endif
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 1.0.0 (2022-01-20)

### Added

- Add option to set a custom `ImageService` [`0fa7799`](https://github.com/area17/twill-image/commit/0fa77994b9489e00d561def79ab72327ad195880)

### Fixed

- Fix bug where cached image are revealed twice [`#33`](https://github.com/area17/twill-image/pull/33) [`ffcdc93`](https://github.com/area17/twill-image/commit/ffcdc93691c1bae4de1566bc77cbff71369d5cad)
- Apply image styles to placeholder and main `img` tag [`b5a6fc2`](https://github.com/area17/twill-image/commit/b5a6fc24eec990da122d16e9cd52b463bc940f93)
- Check image width to set minimum in `srcset` [`233c870`](https://github.com/area17/twill-image/commit/233c8708ed7f370e33be9b48241df59d3298dbf3)
- Check source width to set image default [`253bf74`](https://github.com/area17/twill-image/commit/253bf743438aa5282b5c24bde7227df5748df26b)
- Remove duplicate background color style [`f7edf3e`](https://github.com/area17/twill-image/commit/f7edf3e6ec0ac5d6fee0e63510778734f3bf8a3e)

## 1.0.0-rc4 (2021-11-13)

### Added

- Allow using native lazyload with `loading="lazy"` without the JavaScript module [`3db2ef3`](https://github.com/area17/twill-image/commit/3db2ef3372aecea2c3a3b8028d8e59bdb4219da5)

### Fixed

- Fix WebP srcset not using the right set [`c6e8f0b`](https://github.com/area17/twill-image/commit/c6e8f0b5842fba47c176cb62734e678da242980f)

## 1.0.0-rc3 (2021-11-08)

### Added

- Minor optimizations when webp_support is not enabled [`c3e06fe`](https://github.com/area17/twill-image/commit/c3e06feed527f889c53f6c56d312de311000432c)

### Fixed

- Fix problem with uppercase extension [`6be42f4`](https://github.com/area17/twill-image/commit/6be42f402718d17e1333eabdccf83c192123c198)

## 1.0.0-rc2 (2021-10-05)

### Fixed

- Fix wrong variable name in placehoder template [`f17f279`](https://github.com/area17/twill-image/commit/f17f27996d0a994d0bd14610884d8a3113c1a9a7)

## 1.0.0-rc1 (2021-10-01)

### Added

- Complete refactor of the API
- Fluent image model
- Creation of an agnostic MediaSource service
- Isolate frontend view logic in a view model
- Refactor service provider
- Allow image profiles to be re-used on any image role
- Review profile structure
- Make image role with single crop easy to render without profile
- Add sizes attribute dynamic generation with ImageColumns service

[`fc8b4a0`](https://github.com/area17/twill-image/commit/fc8b4a0), [`d724edf`](https://github.com/area17/twill-image/commit/d724edf), [`7497ad2`](https://github.com/area17/twill-image/commit/7497ad2), [`eec9cd8`](https://github.com/area17/twill-image/commit/eec9cd8), [`1f12e5e`](https://github.com/area17/twill-image/commit/1f12e5e), [`29a6ca1`](https://github.com/area17/twill-image/commit/29a6ca1), [`cb31ca0`](https://github.com/area17/twill-image/commit/cb31ca0), [`9ebc315`](https://github.com/area17/twill-image/commit/9ebc315), [`77c6bcf`](https://github.com/area17/twill-image/commit/77c6bcf), [`3bf71fd`](https://github.com/area17/twill-image/commit/3bf71fd), [`6ba6ab5`](https://github.com/area17/twill-image/commit/6ba6ab5), [`f11d66c`](https://github.com/area17/twill-image/commit/f11d66c), [`44c6391`](https://github.com/area17/twill-image/commit/44c6391), [`56049eb`](https://github.com/area17/twill-image/commit/56049eb), [`85f9070`](https://github.com/area17/twill-image/commit/85f9070), [`3d1d353`](https://github.com/area17/twill-image/commit/3d1d353), [`233ff70`](https://github.com/area17/twill-image/commit/233ff70), [`844e504`](https://github.com/area17/twill-image/commit/844e504), [`d65b5ee`](https://github.com/area17/twill-image/commit/d65b5ee), [`9fb378b`](https://github.com/area17/twill-image/commit/9fb378b), [`57ea081`](https://github.com/area17/twill-image/commit/57ea081), [`203200a`](https://github.com/area17/twill-image/commit/203200a), [`95a0421`](https://github.com/area17/twill-image/commit/95a0421), [`a4c04dd`](https://github.com/area17/twill-image/commit/a4c04dd), [`4fe4a64`](https://github.com/area17/twill-image/commit/4fe4a64), [`03e3d7f`](https://github.com/area17/twill-image/commit/03e3d7f), [`658891b`](https://github.com/area17/twill-image/commit/658891b), [`4cb2433`](https://github.com/area17/twill-image/commit/4cb2433)

- Re-write of the documentation

[`92c5b79`](https://github.com/area17/twill-image/commit/92c5b79), [`8287598`](https://github.com/area17/twill-image/commit/8287598), [`78746cb`](https://github.com/area17/twill-image/commit/78746cb), [`5771dc1`](https://github.com/area17/twill-image/commit/5771dc1), [`f891985`](https://github.com/area17/twill-image/commit/f891985), [`ba45f43`](https://github.com/area17/twill-image/commit/ba45f43), [`3c0d313`](https://github.com/area17/twill-image/commit/3c0d313), [`078a05a`](https://github.com/area17/twill-image/commit/078a05a), [`1d59227`](https://github.com/area17/twill-image/commit/1d59227), [`10a27cf`](https://github.com/area17/twill-image/commit/10a27cf), [`4bbfb04`](https://github.com/area17/twill-image/commit/4bbfb04), [`751a56d`](https://github.com/area17/twill-image/commit/751a56d), [`f418e47`](https://github.com/area17/twill-image/commit/f418e47), [`e33bb7b`](https://github.com/area17/twill-image/commit/e33bb7b)

### Fixed

- Fixed wrong opacity set on `noscript` tag [`4d79ee9`](https://github.com/area17/twill-image/commit/4d79ee9d755346470df89f3c5cd3f692c189ab7f)
- Import ImageService facade [`d2882d0`](https://github.com/area17/twill-image/commit/d2882d0)
- Webp first on main source [`5a78913`](https://github.com/area17/twill-image/commit/5a78913)
- List webp srcset first [`7ae2e30`](https://github.com/area17/twill-image/commit/7ae2e30)
- update path-parse dependency [`80ffd88`](https://github.com/area17/twill-image/commit/80ffd88)

### Improved

- Add error message when role isn't found [`445b7de`](https://github.com/area17/twill-image/commit/445b7de)

### Chores


src/Models/Source.php
<?php

namespace A17\Twill\Image\Models;

use ImageService;
use A17\Twill\Models\Media;
use A17\Twill\Models\Behaviors\HasMedias;
use Illuminate\Contracts\Support\Arrayable;
use A17\Twill\Image\Exceptions\ImageException;
use A17\Twill\Image\Models\Interfaces\ImageSource;

class Source implements Arrayable
{
    public const AUTO_WIDTHS = [
        250,
        500,
        750,
        1000,
        1500,
        2000,
        2500,
        3000,
        3500,
        4000,
        4500,
        5000,
    ];

    public const AUTO_WIDTHS_RATIO = 2.5;

    public const DEFAULT_WIDTH = 1000;

    public const TYPE_GIF = 'image/gif';

    public const TYPE_JPEG = 'image/jpeg';

    public const TYPE_WEBP = 'image/webp';

    protected $model;

    protected $role;

    protected $crop;

    protected $media;

    protected $preset;

    protected $imageArray;

    protected $sources;

    protected $sizes;

    protected $width;

    /**
     * Build a Source to be used with A17\Twill\Image\Models\Image
     *
     * @param [A17\Twill\Models\Model] $object of type Media, Block, module, etc.
     * @param string $role Twill Media role
     * @param array $args Arguments
     * @param string $preset Preset name
     * @param Media $media Twill Media instance
     * @throws ImageException
     */
    public function __construct(
        $object,
        $role,
        $args = [],
        $preset = null,
        Media $media = null
    ) {
        $this->role = $role;
        $this->media = $media;
        $this->setModel($object);
        $this->setPreset($preset);
        $this->setCrop($args['crop'] ?? null);
        $this->setImageArray();
        $this->setSources($args['sources'] ?? null);
        $this->setSizes($args['sizes'] ?? null);
        $this->setWidth($args['width'] ?? null);
    }

    protected function setSources($sources)
    {
        $this->sources = $sources ?? $this->preset['sources'] ?? [
            [
                'crop' => $this->crop,
            ]
        ];
    }

    protected function setSizes($sizes)
    {
        $this->sizes = $sizes ?? $this->preset['sizes'] ?? null;
    }

    protected function setWidth($width)
    {
        $this->width = $width ?? $this->preset['width'] ?? self::DEFAULT_WIDTH;
    }

    /**
     * Provide the text description of the image
     *
     * @return string
     */
    public function alt(): string
    {
        return $this->model->imageAltText($this->role, $this->media);
    }

    /**
     * Provide the caption of the image
     *
     * @return string
     */
    public function caption(): string
    {
        return $this->model->imageCaption($this->role, $this->media);
    }

    public function defaultSrc()
    {
        $defaultWidth = $this->width;

        return $this->model->image(
            $this->role,
            $this->crop,
            ['w' => $defaultWidth],
            false,
            false,
            $this->media,
        );
    }


    public function lqip()
    {
        $sources = [];

        foreach ($this->sources as $source) {
            $crop = $source['crop'] ?? $this->crop;

            $sources[] = [
                'mediaQuery' => $source['media_query'] ?? null,
                'crop' => $crop,
                'type' => self::TYPE_GIF,
                'srcset' => sprintf(
                    '%s 1x',
                    $this->media
                        ? $this->media->pivot->lqip_data ??
                            ImageService::getTransparentFallbackUrl()
                        : $this->model->lowQualityImagePlaceholder(
                            $this->role,
                            $crop,
                        ),
                ),
            ];
        }

        return [
            'src' => $this->media
                ? $this->media->pivot->lqip_data ??
                    ImageService::getTransparentFallbackUrl()
                : $this->model->lowQualityImagePlaceholder(
                    $this->role,
                    $this->crop,
                ),
            'sources' => $sources,
        ];
    }

    public function srcSets()
    {
        $srcSets = [];

        foreach ($this->sources() ?? [] as $sources) {
            $srcset = [];

            foreach ($sources['sources'] as $source) {
                if (isset($source['src']) && isset($source['width'])) {
                    $srcset[] = sprintf(
                        '%s %dw',
                        $source['src'],
                        $source['width'],
                    );
                }
            }

            $srcSets[] = [
                'srcset' => join(',', $srcset),
                'type' => $sources['type'],
                'mediaQuery' => $sources['mediaQuery'],
                'crop' => $sources['crop'],
                'ratio' => $this->ratio($sources['crop']),
            ];
        }

        return $srcSets;
    }

    protected function sources()
    {
        $sources = [];

        // webp
        if (config('twill-image.webp_support')) {
            foreach ($this->sources as $source) {
                $sources[] = [
                    'mediaQuery' => $source['media_query'] ?? null,
                    'type' => self::TYPE_WEBP,
                    'crop' => $source['crop'] ?? $this->crop,
                    'sources' => $this->imageSources($source, ['fm' => 'webp']),
                ];
            }
        }

        // jpeg
        foreach ($this->sources as $source) {
            $sources[] = [
                'mediaQuery' => $source['media_query'] ?? null,
                'type' => self::TYPE_JPEG,
                'crop' => $source['crop'] ?? $this->crop,
                'sources' => $this->imageSources($source, ['fm' => 'jpg']),
            ];
        }

        return $sources;
    }

    protected function imageSources($mediaQueryConfig, $sourceParams = [])
    {
        $widths = $mediaQueryConfig['widths'] ?? $this->defaultWidths();
        $sourcesList = [];

        foreach ($widths as $width) {
            $sourcesList[] = [
                'src' => $this->model->image(
                    $this->role,
                    $mediaQueryConfig['crop'] ?? $this->crop,
                    ['w' => $width] + $sourceParams,
                    false,
                    false,
                    $this->media,
                ),
                'width' => $width,
            ];
        }

        return $sourcesList;
    }

    protected function defaultWidths()
    {
        $defaultWidth = $this->width;

        return array_filter(self::AUTO_WIDTHS, function ($width) use (
            $defaultWidth
        ) {
            return $width <= $defaultWidth * self::AUTO_WIDTHS_RATIO;
        });
    }



    /**
     * Build crops list
     *
     * @return array
     */
    protected function getAllCrops(): array
    {
        $role = $this->role;

        $crops = array_values(
            $this->model->medias
                ->filter(function ($media) use ($role) {
                    return $media->pivot->role === $role;
                })
                ->map(function ($media) {
                    return $media->pivot->crop;
                })
                ->toArray(),
        );

        return $crops;
    }

    /**
     * Set model crop
     *
     * @param null|string $crop
     * @return void
     */
    protected function setCrop($crop)
    {
        if (isset($crop)) {
            $this->crop = $crop;
            return;
        }

        if (isset($this->preset['crop'])) {
            $this->crop = $this->preset['crop'];
            return;
        }

        $crops = $this->getAllCrops();

        if ($index = array_search('default', $crops)) {
            return $crops[$index];
        }

        if (count($crops) > 1) {
            throw new ImageException(
                "Image role has more than one crop, please add 'crop' key to the configuration preset",
                1,
            );
        }

        $this->crop = $crops[0];
    }

    private function extension(): string
    {
        return pathinfo($this->media()->filename, PATHINFO_EXTENSION);
    }

    private function media()
    {
        return $this->media ?? $this->model->imageObject($this->role, $this->crop);
    }

    private function mimeType($extension)
    {
        $types = [
            'png' => 'image/png',
            'jpe' => 'image/jpeg',
            'jpeg' => 'image/jpeg',
            'jpg' => 'image/jpeg',
            'gif' => 'image/gif',
            'bmp' => 'image/bmp',
            'ico' => 'image/vnd.microsoft.icon',
            'tiff' => 'image/tiff',
            'tif' => 'image/tiff',
            'svg' => 'image/svg+xml',
            'svgz' => 'image/svg+xml',
            'webp' => 'image/webp',
        ];

        if (isset($types[$extension])) {
            return $types[$extension];
        }

        return null;
    }

    /**
     * Set model
     *
     * @param object $model
     * @return void
     */
    protected function setModel($model)
    {
        if (!classHasTrait($model, HasMedias::class)) {
            throw new ImageException('Model must use HasMedias trait', 1);
        }

        $this->model = $model;
    }

    protected function setPreset($preset)
    {
        $preset = $preset ?? $this->role;

        if (!config()->has("twill-image.presets.$preset")) {
            $this->preset = null;
            return;
        }

        $this->preset = config("twill-image.presets.$preset");
    }

    protected function setImageArray()
    {
        $imageArray = $this->model->imageAsArray(
            $this->role,
            $this->crop,
            [],
            $this->media,
        );

        if (empty($imageArray)) {
            throw new ImageException(
                "No media was found for role '{$this->role}' and crop '{$this->crop}'",
                1,
            );
        }

        $this->imageArray = $imageArray;
    }

    protected function ratio($crop): string
    {
        $role = $this->role;

        $media = $this->model->medias
            ->filter(function ($media) use ($role, $crop) {
                return $media->pivot->role === $role &&
                    $media->pivot->crop === $crop;
            })
            ->first();

        return $media->pivot->ratio;
    }

    public function toArray()
    {
        return [
            'lqip' => $this->lqip(),
            'sources' => $this->srcSets(),
            'src' => $this->defaultSrc(),
            'width' => $this->imageArray['width'],
            'height' => $this->imageArray['height'],
            'sizes' => $this->sizes,
            'alt' => $this->alt(),
        ];
    }
}
